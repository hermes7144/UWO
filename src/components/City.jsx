import React from 'react';
import { useQuery } from '@tanstack/react-query';
import { getGoods } from '../api/firebase';

export default function City(city) {
  const { data: goods } = useQuery(['goods', city.city], () => getGoods(city.city), { staleTime: Infinity, enabled: !!city.city });
  const { data: nextGoods } = useQuery(['goods', city.nextCity], () => getGoods(city.nextCity), { staleTime: Infinity, enabled: !!city.nextCity });

  const nextItems = nextGoods && nextGoods.map((nextgood) => nextgood.goods_nm);

  return (
    <table>
      <tbody>
        <tr>
          <th>{city.city}</th>

          {goods &&
            goods.map((good, index) => (
              <td key={index} className={nextItems && nextItems.includes(good.goods_nm) ? 'text-red-600' : ''}>
                {good.goods_nm}
              </td>
            ))}
        </tr>
      </tbody>
    </table>
  );
}
