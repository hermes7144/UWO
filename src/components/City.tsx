import { useEffect } from 'react';
import { BsFillTrashFill } from 'react-icons/bs';
import { useCoordinatesContext } from '../context/CoordinatesContext';
import useCity from '../Hooks/useCity';

type CityProps = {
  citys: number[];
  index: number;
  cityNm: string;
  isEditable?: boolean;
  coordinates: [number, number];
  onDelete: (index: number) => void;
};

export default function City({ citys, index, cityNm, coordinates, isEditable = true, onDelete }: CityProps) {
  const { setCoordinates } = useCoordinatesContext();
  const {
    goodsQuery: { isLoading, data: goods },
  } = useCity(citys[index]);
  const {
    goodsQuery: { data: nextGoods },
  } = useCity(citys[index + 1]);

  const nextItems = nextGoods && nextGoods.map((nextgood) => nextgood.goods_nm);
  const handleDelete = () => onDelete(index);

  const handleClick = (coordinates) => setCoordinates(coordinates);
  useEffect(() => {
    index === 0 && setCoordinates(coordinates);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <section>
      {isLoading && <p>Loading...</p>}
      <table className='border-solid border-2 my-2'>
        <tbody>
          <tr>
            <th className='border-solid border-2 w-6' rowSpan={2}>
              {index + 1}
            </th>
            <th rowSpan={2}>
              <div className='flex flex-col justify-center items-center w-24'>
                <button onClick={() => handleClick(coordinates)}>{cityNm}</button>

                {isEditable && <BsFillTrashFill className='cursor-pointer opacity-50 hover:opacity-100' onClick={handleDelete} />}
              </div>
            </th>
            {goods &&
              goods.map((good, index) => (
                <td key={index} className={'border-solid border-2 w-30 ' + (good.specialty ? 'bg-yellow-200' : '')}>
                  <img className='m-auto' src={good.goods_url} alt='' />
                </td>
              ))}
          </tr>
          <tr>
            {goods &&
              goods.map((good, index) => (
                <td key={index} className='border-solid border-2 text-center text-xs'>
                  <span className={nextItems && nextItems.includes(good.goods_nm) ? 'text-red-600' : ''}>{good.goods_nm}</span>
                </td>
              ))}
          </tr>
        </tbody>
      </table>
    </section>
  );
}
