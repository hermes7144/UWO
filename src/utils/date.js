// import { format, register } from 'timeago.js';
// import koLocale from 'timeago.js/lib/lang/ko';

// register('ko', koLocale);

// export function formatAgo(date, lang = 'en_US') {
//   return format(date, lang);
// }

export function formatDate(rawDate) {
  const date = new Date(rawDate);
  const year = date.getFullYear().toString().slice(-2); // 연도에서 뒤의 두 자리만 추출
  const month = ('0' + (date.getMonth() + 1)).slice(-2); // 월을 두 자리 숫자로 변환
  const day = ('0' + date.getDate()).slice(-2); // 일을 두 자리 숫자로 변환
  return `${year}/${month}/${day}`;
}