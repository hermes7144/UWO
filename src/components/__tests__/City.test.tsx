import { render, screen } from '@testing-library/react';
import { Route } from 'react-router-dom';
import { fakeGoods } from '../../tests/fakeGoods';
import { withAllContexts, withRouter } from '../../tests/utils';
import City from '../City';

describe('City', () => {
  const fakeUseCity = jest.fn();

  beforeEach(() => {
    fakeUseCity.mockImplementation(() => {
      return {
        goodsQuery: { isLoading: false, data: fakeGoods },
      };
    });
  });

  afterEach(() => fakeUseCity.mockReset());

  it('renders correctlyd', async () => {
    const { asFragment } = renderCity();

    await screen.findAllByText('아몬드');
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders city name', () => {
    renderCity();
    const cityName = screen.getByText('test city');
    expect(cityName).toBeInTheDocument();
  });

  function renderCity() {
    return render(withAllContexts(withRouter(<Route path='/' element={<City route={{}} index={0} city={{ city_id: 1, city_nm: 'test city', city_coordinates: [37.5665, 126.978] }} />} />), fakeUseCity));
  }
});
