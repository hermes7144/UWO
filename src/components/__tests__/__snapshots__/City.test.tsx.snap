// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`City renders correctly 1`] = `
<DocumentFragment>
  <section>
    <table
      class="border-solid border-2 my-2"
    >
      <tbody>
        <tr>
          <th
            class="border-solid border-2 w-6"
            rowspan="2"
          >
            1
          </th>
          <th
            rowspan="2"
          >
            <div
              class="flex flex-col justify-center items-center w-24"
            >
              <button>
                test city
              </button>
            </div>
          </th>
          <td
            class="border-solid border-2 w-30 "
          >
            <img
              alt=""
              class="m-auto"
              src="http://res.cloudinary.com/da25q6n7o/image/upload/v1678705707/prido8t3x7zzvpmomtoh.png"
            />
          </td>
          <td
            class="border-solid border-2 w-30 bg-yellow-200"
          >
            <img
              alt=""
              class="m-auto"
              src="http://res.cloudinary.com/da25q6n7o/image/upload/v1678706021/ln48swpjtphyvymib58z.png"
            />
          </td>
        </tr>
        <tr>
          <td
            class="border-solid border-2 text-center text-xs"
          >
            <span
              class="text-red-600"
            >
              브랜디
            </span>
          </td>
          <td
            class="border-solid border-2 text-center text-xs"
          >
            <span
              class="text-red-600"
            >
              아몬드
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </section>
</DocumentFragment>
`;
